### ISSUE TITLE
Destroying and Recreating the Globe

### ISSUE THREAD

**Issue by mannodermaus** [glob3mobile/g3m#53]
Hey guys,

I would like to temporarily hide the Globe when my application is showing other Views and it isn't visible anymore. How would I go about re-creating the Globe after it was destroyed? I tried the following (all code snippets are abbreviated):

_1. "Pause and Resume"_
I don't want to initialize a new Globe, so I temporarily remove it from its wrapper...

```
globe.setVisibility(View.GONE);
wrapperLayout.removeView(globe);
globe.(…).invokeInRendererThread(new GTask() {
            @Override
            public void run(G3MContext arg0) {
                globe.onPause();
            }
        }, true);
```

…, and re-attach it at a later point in time:

```
globe.setVisibility(View.VISIBLE);
wrapperLayout.addView(globe);
globe.(…).invokeInRendererThread(new GTask() {
            @Override
            public void run(G3MContext arg0) {
                globe.onResume();
            };
        }, true);
```

This approach fails because of a potential partial re-initialization (please point out if I'm wrong, as this is basically just speculations): The Globe may need to initialize a new OpenGL thread upon regaining its visibility, but the widget used is still linked to the old one. Hence, I'm getting RuntimeExceptions:

```
java.lang.RuntimeException: OpenGL code executed from a Non-OpenGL thread.
(OpenGLThread=Thread[GLThread 5186,5,main],
CurrentThread=Thread[GLThread 5218,5,main])
```

_2. "Total annihilation"_
My next approach was way more brutal: Kill everything related to the Globe and wipe out any trace of the previous Globe before instantiating the new one. This seemed logical to me, and I managed to remove all Globe resources (those that I can think of anyway), passing them to the GC.

```
// Running on the Globe's Renderer thread:
globe.onDestroy();
markerRenderer.dispose();
weatherLayerRenderer.dispose();
globe.getG3MWidget().dispose();

globe = null;
markerRenderer = null;
weatherLayerRenderer = null;
```

A quick check using HPROF heap dumps revealed that I indeed managed to kill all objects sharing the G3M name and MarksRenderer. Upon the next initialization, however, I was greeted with a strange behaviour related to the one mentioned in issue #31: After having created the second Globe instance, all renderers would work as expected but the MarksRenderer. My addMark() operations were running on the Globe's Renderer thread, as requested by the core, however I couldn't seem to make the Markers added to the MarksRenderer appear after I already destroyed the Globe once using the previously mentioned method.

tl;dr: Is there any way to safely destroy and recreate, or hide and show the Globe again and have every component behave as it did before? Any help is much appreciated!

Thanks,
Marcel

**Comment by DiegoGomezDeck** [comment]
Hello Marcel,

The "hack" you tried has to work, and I found a problem on MarkRenderer (it's using a static variable that has opengl state inside and, then, it can't be reused later on a different opengl context).  I'll take a look to try to cleanup it when the opengl context gets removed.

I'll keep you informed.

-- Diego

**Comment by DiegoGomezDeck** [comment]
Hi Marcel,

I just pushed a new version of MarksRenderer that it's supposed to fix this problem.  I didn't test it, can you give it a try and inform me if this works?

TIA,

-- Diego

**Comment by mannodermaus** [comment]
Hey Diego,

using the second approach, I managed to have the markers appear with subsequent rendering contexts as well! I will look into the first option later. Thanks so much!

Marcel

**Comment by mannodermaus** [comment]
Hey Diego,

the first approach (removeView() and addView()) still fails with a RuntimeException stating that the "new" GLThread isn't the one the globe used to know (which makes sense). Just wanted to let you know that this still happens, however I'm sticking to the other approach anyways since it seems much cleaner to me from a GC standpoint. If you'd want to work on this issue some more and would like me to test it, I would gladly do so. If not, there is one working solution. Thanks again!

Marcel
